### PR TITLE
More docs for HttpUrl.

### PR DESCRIPTION
**Pull request by swankjesse**
Closes: https://github.com/square/okhttp/issues/2577
Closes: https://github.com/square/okhttp/issues/2263

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 1]
```diff
@@ -470,6 +550,17 @@ static void pathSegmentsToString(StringBuilder out, List<String> pathSegments) {
     }
   }
 
+  /**
+   * Returns a a list of encoded path segments like {@code ["a", "b", "c"]} for the URL {@code
```
Comment (dave-r12) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: There's an extra 'a' here.

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 2]
```diff
@@ -571,6 +734,23 @@ public String queryParameter(String name) {
     return Collections.unmodifiableSet(result);
   }
 
+  /**
+   * Returns all values for the query parameter {@code name} ordered by their appearance in this
+   * URL. For example this returns {@code ["banana"]} for {@code queryParameterValue("b")} on {@code
+   * http://host/?a=apple&b=banana}.
+   *
+   * <p><table summary="">
+   *   <tr><th>URL</th><th>queryParameterValues("a")</th><th>queryParameterValues("b")</th></tr>
```
Comment (dave-r12) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: These examples are really helpful.

**Comment by dave-r12** [comment]
Lgtm.

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 3]
```diff
@@ -386,27 +392,68 @@ public boolean isHttps() {
     return scheme.equals("https");
   }
 
-  /** Returns the username, or an empty string if none is set. */
+  /**
+   * Returns the username, or an empty string if none is set.
+   *
+   * <p><table summary="">
+   *   <tr><th>URL</th><th>encodedUsername()</th></tr>
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: Surround `encodedUsername()` with `{@code ..}`

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 4]
```diff
@@ -386,27 +392,68 @@ public boolean isHttps() {
...
     int usernameStart = scheme.length() + 3; // "://".length() == 3.
     int usernameEnd = delimiterOffset(url, usernameStart, url.length(), ":@");
     return url.substring(usernameStart, usernameEnd);
   }
 
+  /**
+   * Returns the decoded username, or an empty string if none is present.
+   *
+   * <p><table summary="">
+   *   <tr><th>URL</th><th>username()</th></tr>
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: `{@code ..}`

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 5]
```diff
@@ -386,27 +392,68 @@ public boolean isHttps() {
...
   public String username() {
     return username;
   }
 
-  /** Returns the password, or an empty string if none is set. */
+  /**
+   * Returns the password, or an empty string if none is set.
+   *
+   * <p><table summary="">
+   *   <tr><th>URL</th><th>encodedPassword()</th></tr>
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: `{@code ..}`

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 6]
```diff
@@ -386,27 +392,68 @@ public boolean isHttps() {
...
     int passwordEnd = url.indexOf('@');
     return url.substring(passwordStart, passwordEnd);
   }
 
-  /** Returns the decoded password, or an empty string if none is present. */
+  /**
+   * Returns the decoded password, or an empty string if none is present.
+   *
+   * <p><table summary="">
+   *   <tr><th>URL</th><th>password()</th></tr>
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: `{@code ..}`

File: okhttp/src/main/java/okhttp3/HttpUrl.java [hunk 7]
```diff
@@ -421,6 +468,14 @@ public String password() {
    *   <li>An IPv6 address, like {@code ::1}. Note that there are no square braces.
    *   <li>An encoded IDN, like {@code xn--n3h.net}.
    * </ul>
+   *
+   * <p><table summary="">
+   *   <tr><th>URL</th><th>host()</th></tr>
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: etc..

**Comment by JakeWharton** [comment]
just one formatting nit. lgtm otherwise.

File: okhttp/src/main/java/okhttp3/HttpUrl.java (hunk 3 above)
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/HttpUrl.java]: Fixed all!

**Pull request merged**